### PR TITLE
Enforce allowable bucket name characters at bucket creation time

### PR DESCRIPTION
**Pull request by kellymclaughlin**
Ensure that bucket names follow the limitations so that they are DNS compliant. The complete list of restrictions is [here](http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html). 

Update all riak_test modules to use valid bucket names for testing. 

Also define a function clause to handle `405` error responses in the webmachine error handler to better conform to the S3 response format. There is a related webmachine pull request [here](https://github.com/basho/webmachine/pull/132) to allow the default behavior of webmachine to be changed so that `405` errors can be handled by the defined webmachine error handler module.

This pull request resolves the following ceph s3-test suite failures:
- s3tests.functional.test_s3.test_bucket_create_naming_bad_starts_nonalpha
- s3tests.functional.test_s3.test_bucket_create_naming_bad_short_empty
- s3tests.functional.test_s3.test_bucket_create_naming_bad_short_one
- s3tests.functional.test_s3.test_bucket_create_naming_bad_short_two
- s3tests.functional.test_s3.test_bucket_create_naming_bad_long
- s3tests.functional.test_s3.test_bucket_create_naming_bad_ip
- s3tests.functional.test_s3.test_bucket_create_naming_bad_punctuation

### CODE REVIEW

File: src/riak_cs_utils.erl [hunk 1]
```diff
@@ -142,13 +142,80 @@ close_riak_connection(Pool, Pid) ->
...
+            ok
+    end.
+
+-spec valid_bucket_name(binary()) -> boolean().
+valid_bucket_name(Bucket) when byte_size(Bucket) < 3 orelse
+                               byte_size(Bucket) > 63 ->
+    false;
+valid_bucket_name(Bucket) ->
+    lists:all(fun(X) -> X end, [valid_bucket_label(Label) ||
+                                   Label <- binary:split(Bucket, [<<".">>])])
```
Comment (reiddraper) [inline src/riak_cs_utils.erl]: `binary:split/2` only splits the first occurrence, which makes this fail with `false`: `valid_bucket_name(<<"this.should.pass">>).`

Comment (reiddraper) [inline src/riak_cs_utils.erl]: Looks like adding the `[global]` option to `binary:split` would fix this.

**Comment by reiddraper** [comment]
Started writing a test suite, on `bugfix/wrd-bucket-name-tests`. A few of them fail for things like `binary:first(<<>>)`, when calling `valid_bucket_name(<<".valid-name">>).`

**Comment by reiddraper** [comment]
Pushed some fixes to the aforementioned branch as well.

**Comment by kellymclaughlin** [comment]
Taking a look at your branch now. Probably going to make a few changes and merge those commits to this branch and we can continue the review and collaboration from here.

File: src/riak_cs_utils.erl [hunk 2]
```diff
@@ -142,13 +142,80 @@ close_riak_connection(Pool, Pid) ->
...
+-spec valid_bucket_char(integer()) -> boolean().
+valid_bucket_char(Char) ->
+    numeric_char(Char) orelse
+        lower_case_char(Char) orelse
+        upper_case_char(Char) orelse
+        dash_char(Char).
+
+-spec numeric_char(integer()) -> boolean().
+numeric_char(Char) ->
+    Char >= 48 andalso Char =< 57.
```
Comment (shino) [inline src/riak_cs_utils.erl]: I like `$0` instead of 48 for readability. (sorry for nitpick :P )

**Comment by kellymclaughlin** [comment]
@shino good suggestion, pushed that change

**Comment by kellymclaughlin** [comment]
I merged in the commits from `bugfix/wrd-bucket-name-tests`. I also made a fix to exclude upper-case characters and added a few more test bucket names.

File: rel/files/app.config [hunk 3]
```diff
@@ -154,7 +154,11 @@
                {log_handlers, [
                                {webmachine_log_handler, ["{{platform_log_dir}}"]},
                                {riak_cs_access_log_handler, []}
-                              ]}
+                              ]},
+
+               %% Allow the Riak CS error handler to define the
+               %% response body for 405 errors
+               {error_response_on_405, true}
```
Comment (shino) [inline rel/files/app.config]: Unused configuration?

File: src/riak_cs_wm_error_handler.erl [hunk 4]
```diff
@@ -20,6 +21,16 @@ render_error(500, Req, Reason) ->
     ErrorFour = <<"this request</body></html>">>,
     IOList = [ErrorOne, ErrorTwo, ErrorThree, ErrorFour],
     {erlang:iolist_to_binary(IOList), ReqState};
+render_error(405, Req, Reason) ->
```
Comment (shino) [inline src/riak_cs_wm_error_handler.erl]: How can I see this response?
I tried some wrong methods but responses have no body xml.

Example of trial:

```
$ s3curl.pl --id local_rcs \
>     -- -s -v -x localhost:8080 \
>     -X POST \
>     http://test-tmp.s3.amazonaws.com/ \
>
> POST http://test-tmp.s3.amazonaws.com/ HTTP/1.1
> User-Agent: curl/7.27.0
> Host: test-tmp.s3.amazonaws.com
> Accept: */*
> Proxy-Connection: Keep-Alive
> Date: Wed, 13 Mar 2013 13:44:19 +0000
> Authorization: AWS VAARM7GQ2CGYX1FPQTBE:/B+Tv3Xl+gMeSmTbqh13t3ckq6E=
>
< HTTP/1.1 405 Method Not Allowed
< Server: Riak CS
< Date: Wed, 13 Mar 2013 13:44:19 GMT
< Content-Length: 0
< Allow: HEAD, PUT, DELETE
<
```

Comment (shino) [inline src/riak_cs_wm_error_handler.erl]: Sorry. I missed the pr of webmachine.

File: rel/files/app.config (hunk 3 above)
Comment (shino) [inline rel/files/app.config]: Also this is related to webmachine... sorry for garbage comment.

**Comment by reiddraper** [comment]
> I merged in the commits from bugfix/wrd-bucket-name-tests. I also made a fix to exclude upper-case characters and added a few more test bucket names.

Excellent. All the tests pass for me. Will do a little more review when I get into the office and then I'm +1.

**Comment by shino** [comment]
Tests pass and works good!

Only one point I want to discuss on webmachine pr. Comment is here: https://github.com/basho/webmachine/pull/132#issuecomment-14843720

**Comment by reiddraper** [comment]
+1 on the `valid_bucket_name` stuff. +1 overall once we resolve the WM pr.

**Comment by reiddraper** [comment]
Excellent, +1 to merge.

**Pull request merged**